### PR TITLE
Adds LastContent function

### PR DESCRIPTION
**Pull request by catenacyber**
For performance reason : no need to rebuild a big slice when we only need the last element

Goes from 30 seconds to 3 seconds to parse oss-fuzz timeout from https://oss-fuzz.com/testcase-detail/5177927826407424 (1Mb file with lots repetitive `option`)

### CODE REVIEW

**Comment by catenacyber** [comment]
Side thought : Another optimization could be to use buffered channels in lex.go

File: rule.go [hunk 1]
```diff
@@ -491,6 +491,15 @@ func (r *Rule) Contents() []*Content {
 	return cs
 }
 
+func (r *Rule) LastContent() *Content {
```
Comment (duanehoward) [inline rule.go]: can you add a test for this please?

Comment (catenacyber) [reply rule.go]: Just pushed, tried to keep your test style

File: rule_test.go [hunk 2]
```diff
@@ -1038,6 +1038,30 @@ func TestRE(t *testing.T) {
 	}
 }
 
+func TestLastContent(t *testing.T) {
+	for _, tt := range []struct {
+		rule string
+		want string
```
Comment (duanehoward) [inline rule_test.go]: This test should probably not be dependent on the stringer implementation and should include the ContentOptions that we'd expect as well.

```
  for ... []struct {
    rule string
    want *Content
  }{
  ...
    rule = `...`
    want = &Content{...} 
  }

File: rule_test.go [hunk 3]
```diff
@@ -1038,6 +1038,30 @@ func TestRE(t *testing.T) {
...
+	}{
+		{
+			rule: `alert udp $HOME_NET any -> $EXTERNAL_NET any (sid:1337; msg:"foo"; content:"foo"; content:"bar"; within:40;)`,
+			want: `bar`,
+		},
+		{
+			rule: `alert udp $HOME_NET any -> $EXTERNAL_NET any (sid:1337; msg:"foo"; content:"bar"; content:"foo"; within:40;)`,
```
Comment (duanehoward) [inline rule_test.go]: can you add an additional test where the OrderedMatchers contain non-content types as well (PCRE, ByteMatch, etc.) to ensure that we're really returning the last Content and not something else.

Comment (catenacyber) [reply rule_test.go]: Done and added a test case for the nil return value as well

File: rule_test.go (hunk 2 above)
Comment (catenacyber) [reply rule_test.go]: Ok with `reflect.DeepEqual`

File: rule_test.go [hunk 4]
```diff
@@ -1038,6 +1038,42 @@ func TestRE(t *testing.T) {
...
+		{
+			rule: `alert udp $HOME_NET any -> $EXTERNAL_NET any (sid:1337; msg:"foo";)`,
+			want: nil,
+		},
+	} {
+		r, err := ParseRule(tt.rule)
+		if err != nil {
+			t.Fatalf("re: parse rule failed: %v", err)
+		}
+		if got := r.LastContent(); !reflect.DeepEqual(got, tt.want) {
```
Comment (duanehoward) [inline rule_test.go]: Looks great, last nit the other test files moved from reflect.DeepEqual to using "github.com/kylelemons/godebug/pretty" as it makes diffing in tests much more readable.
Can you do the same here? You can find examples in https://github.com/google/gonids/blob/master/optimize_test.go

Comment (catenacyber) [reply rule_test.go]: This package seems nice indeed :-)

**Comment by duanehoward** [comment]
> Side thought : Another optimization could be to use buffered channels in lex.go

@catenacyber Quick random question, was there a specific reason you chose the `0x1000` size for the buffered channel in `lex.go`?

**Comment by catenacyber** [comment]
> Quick random question, was there a specific reason you chose the 0x1000 size for the buffered channel in lex.go?

Nothing specific : this is the usual size of a memory page, so what ever the size of the object, we will fit precisely into a number of pages, and so it should be optimized...

What do you think ?

**Comment by duanehoward** [comment]
> > Quick random question, was there a specific reason you chose the 0x1000 size for the buffered channel in lex.go?
> 
> Nothing specific : this is the usual size of a memory page, so what ever the size of the object, we will fit precisely into a number of pages, and so it should be optimized...
> 
> What do you think ?

All good! I had a colleague ask about this "magic number" and I hadn't asked during review. Thanks for the update!

**Comment by catenacyber** [comment]
>  I had a colleague ask about this "magic number

This is indeed pretty magic and not fully tested  :it is better than 1, but is it better than 256 ?
There may well be a better choice...

**Pull request merged**